s3prl/downstream/speech_translation/S3prl_SpeechToTextTask.py

```diff
@@ -0,0 +1,211 @@
+            
+            # sample rate
+            srs.extend([int(ss.get(cls.KEY_SAMPLE_RATE, cls.DEFAULT_SAMPLE_RATE)) for ss in s])
+        
+        return S3prl_SpeechToTextDataset(
+            split_name,
+            is_train_split,
+            data_cfg,
+            audio_paths,
+            n_frames,
+            src_texts = src_texts,
+            tgt_texts = tgt_texts,
+            speakers = speakers,
+            src_langs = src_langs,
+            tgt_langs = tgt_langs,
+            srs = srs,
+            ids = ids,
+            tgt_dict = tgt_dict,
+            pre_tokenizer = pre_tokenizer,
+            bpe_tokenizer = bpe_tokenizer,
+            upstream_rate = upstream_rate,
+        )
+
+    @classmethod
+    def from_tsv(
+        cls,
+        root: str,
+        data_cfg: S2TDataConfig,
+        splits: str,
+        tgt_dict,
+        pre_tokenizer,
+        bpe_tokenizer,
+        is_train_split: bool,
+        epoch: int,
+        seed: int,
+        upstream_rate: int,
+    ) -> SpeechToTextDataset:
+
+        _splits = splits.split(",")
+        assert len(_splits) == 1, "do not support multiple files training"
+
+        split = _splits[0]
+        tsv_path = op.join(root, f"{split}.tsv")
+        if not op.isfile(tsv_path):
+            raise FileNotFoundError(f"Dataset not found: {tsv_path}")
+        
+        with open(tsv_path) as f:
+            reader = csv.DictReader(
+                f,
+                delimiter="\t",
+                quotechar=None,
+                doublequote=False,
+                lineterminator="\n",
+                quoting=csv.QUOTE_NONE,
+            )
+            samples = [dict(e) for e in reader]
+
+        assert len(samples) > 0
+
+        return cls._from_list(
+            split,
+            is_train_split,
+            [samples],
+            data_cfg,
+            tgt_dict,
+            pre_tokenizer,
+            bpe_tokenizer,
+            upstream_rate,
+        )
+
+
+class S3prl_SpeechToTextDataset(SpeechToTextDataset):
+
+    TARGET_RATE = 16000
+
+    def __init__(self, *args, srs = Optional[List[int]], upstream_rate = 160, **kwargs):
+        
+        super().__init__(*args, **kwargs)
+
+        self.srs = srs
+
+        self.resamplers = {}
+        for sr in set(srs):
+            self.resamplers[sr] = torchaudio.transforms.Resample(
+                orig_freq = sr,
+                new_freq = self.TARGET_RATE,
+            )
+
+        for i in range(len(self.n_frames)):
+            new_n_frames = self.n_frames[i] * self.TARGET_RATE / self.srs[i] / upstream_rate
+            self.n_frames[i] = int(new_n_frames)
+
+    def __getitem__(
+        self, index: int
+    ) -> Tuple[int, torch.Tensor, Optional[torch.Tensor]]:
+
+        source, sr = torchaudio.load(self.audio_paths[index])
+
+        assert self.srs[index] == sr
+        source = self.resamplers[sr](source)
+        source = torch.mean(source, dim=0)
```

Comment:
I'm not sure if this is OK for multi-channel data. In many cases, just use one channel would be fine. I'm not sure if the mean operation will have some side effects, let's check with others.

Reply:
@ftshijt may I make sure that you means we could just select one channel (e.g. the first channel) and discard other channels?

BTW, @leo19941227 does s3prl support use multichannel waveform as input? if not, is there a consistence way to covert the multichannel data into mono in s3prl?

Reply:
I think we will have a consistent way to do this, specifically for hidden set. I have tested mean & first channel on ASR and the difference is quite trivial on the hidden data. Actually I will preprocessed the audio consistently and distribute to each task in the single channel format. So we can first ignore this issue for now and the way you load waveform will be different eventually.